tests_e2e.py

```diff
@@ -0,0 +1,780 @@
+
+        # Store the user & also the verify token for use in upcoming tests
+        USERS[username] = body
+
+        USERS[username]['verify_token'] = response['body']['token']
+        return USERS[username]
+
+    # Returns the first created user, if any; otherwise, creates one.
+    def get_any_user(self):
+        if len(USERS.keys()) > 0:
+            return USERS[next(iter(USERS))]
+        return self.assert_user_exists(self.make_username())
+
+    # Returns the first logged in user, if any; otherwise, logs in a user; if no user exists, creates and then logs in the user.
+    def get_any_logged_user(self):
+        for user in USERS:
+            if 'cookie' in user:
+                return user
+
+        # If there's no logged in user, we login the user
+        user = self.get_any_user()
+        return self.login_user(user['username'])
+
+    def login_user(self, username):
+        user = USERS[username]
+        if 'cookie' in user:
+            return user
+        response = request('post', 'auth/login', {}, {'username': user['username'], 'password': user['password']})
+        cookie = self.get_hedy_cookie(response['headers']['Set-Cookie'])
+
+        # The cookie value must be set to `hedy={{SESSION}};` so that it can be used as a Cookie header in subsequent requests
+        USERS[user['username']]['cookie'] = CONFIG['session']['cookie_name'] + '=' + cookie.value + ';'
+        return user
+
+    def assert_user_is_logged(self, username):
+        self.assert_user_exists(username)
+        return self.login_user(username)
+
+    def get_hedy_cookie(self, cookie_string):
+        cookie = SimpleCookie()
+        cookie.load(cookie_string)
+
+        for key, cookie in cookie.items():
+            if key == CONFIG['session']['cookie_name']:
+                return cookie
+
+    def given_fresh_user_is_logged_in(self):
+        username = self.make_username()
+        self.user = self.assert_user_is_logged(username)
+        self.username = username
+        if 'cookie' in self.user:
+            self.cookie = self.user['cookie']
+        else:
+            self.cookie = None
+
+    def given_user_is_logged_in(self):
+        self.user = self.get_any_logged_user()
+        self.username = self.user['username']
+        if 'cookie' in self.user:
+            self.cookie = self.user['cookie']
+        else:
+            self.cookie = None
+
+    def given_any_user(self):
+        self.user = self.get_any_user()
+        self.username = self.user['username']
+        if 'cookie' in self.user:
+            self.cookie = self.user['cookie']
+        else:
+            self.cookie = None
+
+    def post_data(self, path, body, expect_http_code=200, no_cookie=False, return_headers=False):
+        headers = {}
+        if no_cookie == False and hasattr(self, 'cookie') and self.cookie:
+            headers['cookie'] = self.cookie
+
+        response = request('post', path, headers, body)
```

Comment:
Should we not retrieve the cookie from the response as well and set it on `self.cookie` every time if applicable?

You are now assuming that cookies are ONLY set by the http calls you expect to set cookies (login calls etc). If we change the server in the future, those assumptions might not be true anymore, and the tests would not behave the same as a real browser would anymore (and so they wouldn't be testing the right thing).

In fact, this already happens today: the quiz pages liberally change the session variables, which get stored in a cookie. So if we don't constantly update the cookies anytime the server says it wants to, we wouldn't be able to test the quiz properly.

`requests` has cookie jar helpers that make this trivial to do, see the previous iteration of the e2e tests where I made the same change already.